### ISSUE TITLE
✨ Webpack 4 upgrade

### ISSUE THREAD

**Issue by pi0** [nuxt/nuxt.js#2686]
Webpack 4 beta just released and no features/breaking-changes planned for final release. The hardest part for the upgrade is plugins that need to adapt to new hooks system. We have ~1 month time remaining to the final release of webpack and (hopefully) until then all of the webpack plugins used by nuxt.js will be compatible with 4.x.

Official release notes: [v4.0.0-beta.0](https://github.com/webpack/webpack/releases/tag/v4.0.0-beta.0)

### Nuxt.js branch

[TODO]

### Testing version

[TODO]

### Conflicts, nuxt release plains

[TODO]

### Tasks list

[TODO]

**Comment by tonyrewin** [comment]
Maybe is a good moment to rethink about plugins system basis. Most of the time I have this in my nuxt.config.js: `
  plugins: [
    { src: '~plugins/nossr', ssr: false }
  ]
` and just one nossr.js file for all of them. But I did not really check  ssr-ok-plugins.

**Comment by pSchaub** [comment]
Webpack 4 is released now. Let's upgrade!

**Comment by pi0** [comment]
@pSchaub We are on it ;)

**Comment by clarkdo** [comment]
@pi0 
ModuleConcatination  issue has been fixed in https://github.com/webpack-contrib/extract-text-webpack-plugin/pull/728

**Comment by ekoeryanto** [comment]
it is already shipped on latest version?

**Comment by pi0** [comment]
Only nuxt-edge. Will be shipped with nuxt 2

**Comment by ekoeryanto** [comment]
@pi0 thanks, I tried and only `vendor` is deprecated there. but, there is any documentation to use the nuxt-edge?

**Comment by pi0** [comment]
https://medium.com/nuxt/nuxt-2-is-coming-oh-yeah-212c1a9e1a67

**Comment by ekoeryanto** [comment]
thanks, migrating now.

**Comment by ekoeryanto** [comment]
```js
compiler.hooks.compilation.tap('NetlifyServerPushPlugin', compilation => {
      compilation.hooks.htmlWebpackPluginBeforeHtmlGeneration.tap(
        'NetlifyServerPushPlugin',
        data => {
          const assets = data.assets;
          const source = this.generateAssetHeaders(assets);
          compilation.assets[`${this.options.headersFile}`] = {
            source: () => source,
            size: () => source.length,
          };
        }
      );
    });
```
i have a problem, `htmlWebpackPluginBeforeHtmlGeneration` is not exists on webpack. implementing above code not working on nuxt-edge

**Comment by clarkdo** [comment]
@ekoeryanto 
How about to use `compilation.hooks.htmlWebpackPluginBeforeHtmlProcessing.tapAsync`

https://github.com/jantimon/html-webpack-plugin/issues/871#issuecomment-369585543

**Comment by clarkdo** [comment]
Did you install `html-webpack-plugin` in your own package.json and which version are you using?

`htmlWebpackPluginBeforeHtmlProcessing` was introduced since 3.0.0 https://github.com/jantimon/html-webpack-plugin/blob/v3.0.0/index.js#L58

**Comment by ekoeryanto** [comment]
no, I did not install `html-webpack-plugin` my self as it is already covered by nuxt

**Comment by clarkdo** [comment]
@ekoeryanto Can you please provide a repo to reproduce you issue?

**Comment by ekoeryanto** [comment]
@clarkdo here https://github.com/pakeweb/apek/tree/develop
I commented out plugin section in nuxt.config.js

**Comment by clarkdo** [comment]
@ekoeryanto 
Sorry for the late reply, I'm kind of busy with my personal stuff recent days. 
Could you please open an issue and cc to me in case that I won't forget it.

**Comment by ekoeryanto** [comment]
done #3254

**Comment by david-j-davis** [comment]
On the Nuxt website it says version 2.0, but build vendor is still on the site, a bit misleading isn't it? https://nuxtjs.org/api/configuration-build/#vendor. Well now that I am using Nuxt@2.0.0, I have removed vendor, but unfortunately I do not know where I am to place my local vendor js so it builds on my page. I suggest the documentation say that vendor is deprecated with a suggestion on what to use instead.

**Comment by david-j-davis** [comment]
Thanks @manniL, I'll look into it, and do we need to install the webpack package as well or is that bundled already in nuxt?

**Comment by manniL** [comment]
@david-j-davis No, you don't need to install that package. It already comes (with a decent configuration for most use cases) with Nuxt 2 ☺️ )

**Comment by david-j-davis** [comment]
Ok thanks @manniL !
